src/vs/base/common/async.ts

```diff
@@ -1225,11 +1226,19 @@ export class ThrottledWorker<T> extends Disposable {
 			this.pendingWork.push(unit);
 		}
 
+		const timeSinceLastExecution = Date.now() - this.lastExecutionTime;
+
 		// If not throttled, start working directly
 		// Otherwise, when the throttle delay has
 		// past, pending work will be worked again.
-		if (!this.throttler.value) {
+		if (!this.throttler.value && timeSinceLastExecution >= this.options.throttleDelay) {
 			this.doWork();
+		} else {
+			this.throttler.value = new RunOnceScheduler(() => {
+				this.throttler.clear();
+				this.doWork();
+			}, Math.max(this.options.throttleDelay - timeSinceLastExecution, 0));
+			this.throttler.value.schedule();
 		}
```

Comment:
I think the condition here is dangerous where we may end up constantly scheduling a `RunOnceScheduler` if a spammy worker calls `work` with a timeout of `0` always. In other words, if someone manages to call `work` all the time before the scheduler was able to execute, the scheduler gets always re-created and never scheduled.

Reply:
This is a good point, I've made a check to make sure that if there is already a scheduled run in the throttler we don't schedule another one. Instead it will just be picked up by the next doWork call